lib/LANraragi/Plugin/Metadata/MEMS.pm

```diff
@@ -0,0 +1,99 @@
+            { type => 'bool', desc =>  'Use ExHentai link for source instead of E-Hentai link' }
+        ]
+    );
+}
+
+# Mandatory function implemented by every plugin.
+sub get_tags {
+
+    shift;
+    my $lrr_info      = shift;
+    my $logger        = get_logger( 'Mayriad\'s EH Master Script', 'plugins' );
+    my $gallery_id    = '';
+    my $gallery_token = '';
+    my ( $save_jpn_title, $save_additional_metadata, $use_exhentai ) = @_;
+
+    # Use the URL from oneshot parameters or source tag first when applicable.
+    if ( $lrr_info->{oneshot_param} =~ /e(?:x|-)hentai\.org\/g\/(\d+)\/([0-9a-z]+)/i ) {
+        $gallery_id    = $1;
+        $gallery_token = $2;
+        $logger->debug("Directly using gallery ID $gallery_id and token $gallery_token from oneshot parameters.");
+    } elsif ( $lrr_info->{existing_tags} =~ /source:e(?:x|-)hentai\.org\/g\/(\d+)\/([0-9a-z]+)/i ) {
+        $gallery_id    = $1;
+        $gallery_token = $2;
+        $logger->debug("Directly using gallery ID $gallery_id and token $gallery_token from source tag.");
+    } else {
+        # Use the gallery ID and token in the filename to directly locate the gallery. Note that the regex does not have
+        # "$" at the end, so the filename can have other information attached after the identifiers.
+        ( $gallery_id, $gallery_token ) = ( $lrr_info->{archive_title} =~ /.+? \[GID (\d+) GT ([0-9a-z]+)\]/ );
+        if ( $gallery_id eq '' || $gallery_token eq '' ) {
+            my $file_error = 'Skipping archive without connecting to EH, because the archive title does not have valid '
+              . 'gallery identifiers from Mayriad\'s EH Master Script.';
+            $logger->error($file_error);
+            return ( error => $file_error );
+        }
+    }
+
+    # Retrieve metadata directly using EH API.
+    $logger->info('Source identified. Calling E-Hentai metadata plugin to retrieve metadata from EH API.');
+    my ( $eh_all_tags, $eh_title ) = LANraragi::Plugin::Metadata::EHentai::get_tags_from_EH( $gallery_id,
+      $gallery_token, $save_jpn_title, $save_additional_metadata );
+    my %metadata = ( tags => $eh_all_tags );
+
+    # Add source URL and title if possible.
+    if ( $metadata{tags} ne "" ) {
+        # Add the source tag outside get_tags_from_EH(), so that this tag is only added when metadata has been
+        # successfully retrieved; otherwise $metadata{tags} may only contain this source tag and truly untagged
+        # galleries may be incorrectly hidden.
+        my $host         = ( $use_exhentai ? 'exhentai.org' : 'e-hentai.org' );
+        $metadata{tags} .= ", source:$host/g/$gallery_id/$gallery_token";
+        # Title is always updated to hide the identifiers and also to reflect title changes due to rename petitions.
+        $metadata{title} = $eh_title;
+    }
+
+    # Return a hash containing the new metadata to be added to LRR.
+    return %metadata;

```

Comment:
It is good. The plugin used to check `if ( $eh_all_tags ne "" )` and give an error message in another function before it is replaced by the stock plugin call. I did not think that check was very necessary though, because EH API would have returned an error first, which would have already given an error in LANraragi.

By the way, do you not also need a `$logger->info("No matching EH Gallery Found!");` here?